### PR TITLE
i276: Clean up InlineAutocompleteEditText; disable autocomplete on Fire TV Remote app

### PR DESCRIPTION
**Pull request by mcomella**
I haven't had a chance to test the iOS Fire TV Remote App yet – the batteries are dead on the office iPhone. :(

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/widget/InlineAutocompleteEditText.java [hunk 1]
```diff
@@ -463,12 +476,26 @@ private boolean removeAutocompleteOnComposing(final CharSequence text) {
...
                 }
                 return super.commitText(text, newCursorPosition);
             }
 
+            private boolean isCommitTextFromRemoteAppOrSoftKeyboardClear(final CharSequence text) {
+                // Two events call this with text as the empty string: Clear from the soft keyboard
+                // and input from the remote app (which calls this with the empty string to
+                // clear the input, then calls this with the full input). Since we don't want
+                // autocomplete for the remote app and autocomplete is unnecessary when all the text
+                // is cleared, we compare against the empty string to see if this could be the
```
Comment (liuche) [inline app/src/main/java/org/mozilla/focus/widget/InlineAutocompleteEditText.java]: Hm, clever hack!

File: app/src/main/java/org/mozilla/focus/widget/InlineAutocompleteEditText.java [hunk 2]
```diff
@@ -640,18 +613,28 @@ public boolean onKeyPreIme(int keyCode, KeyEvent event) {
...
+            final int start = text.getSpanStart(AUTOCOMPLETE_SPAN);
+
+            if (mSettingAutoComplete || start < 0 || (start == selStart && start == selEnd)) {
+                // Do not commit autocomplete text if there is no autocomplete text
+                // or if selection is still at start of autocomplete text
+                return;
+            }
+
+            if (selStart <= start && selEnd <= start) {
+                // The cursor is in user-typed text; remove any autocomplete text.
```
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/java/org/mozilla/focus/widget/InlineAutocompleteEditText.java]: I think we do highly custom behavior (there's a lot of this autocomplete code!) – e.g. how we highlight the autocomplete on the end of the word, update it when a letter is typed, or remove it when backspace is pressed. Android has some built-in widgets but this is a lot slicker.

File: app/src/main/java/org/mozilla/focus/widget/InlineAutocompleteEditText.java [hunk 3]
```diff
@@ -137,10 +132,6 @@ void setOnSearchStateChangeListener(OnSearchStateChangeListener listener) {
         mSearchStateChangeListener = listener;
     }
 
-    public void setOnTextChangeListener(OnTextChangeListener listener) {
```
[the remark this answers is not in the document]
Comment (mcomella) [reply app/src/main/java/org/mozilla/focus/widget/InlineAutocompleteEditText.java]: Another reason I didn't like this listeners is that they were redundant (and sometimes used differently): we call `addTextChangedListener(new TextChangeListener()` in `onAttachedToWindow`, which is independent of this public `TextChangeListener` method. I guess the idea is we listen for text changed events and relay them to this external TextChangeListener (and how exactly was that supposed to interact with the `onFilterListener`? 😭

**Comment by mcomella** [comment]
> How does this work if you're using both remotes (or is that possible)?

In my testing, it worked seamlessly - autocomplete will turn off if you start typing with the remote app and it will turn back on when you type with the hardware remote.

FYI: will test with the iOS remote app when I get into the office later today. 🖖

**Comment by mcomella** [comment]
I'm just going to land this and I can test the iOS remote app when I'm back in the office. Better to get this into the Live APp test build.

**Pull request merged**